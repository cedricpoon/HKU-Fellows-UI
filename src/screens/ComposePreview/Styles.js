import { StyleSheet } from 'react-native';

import { logo } from 'hkufui/theme/palette';

export default StyleSheet.create({
  send: {
    color: logo.red
  }
});
