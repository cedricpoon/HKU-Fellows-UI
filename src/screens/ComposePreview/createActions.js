import { Linking } from 'react-native';
import { REQUEST_POST_CREATE, END_POST_CREATE } from 'hkufui/src/constants/actionTypes';
import { link, compose, reply } from 'hkufui/config/webapi';
import { localize } from 'hkufui/locale';
import { LOADING, STILL } from 'hkufui/src/constants/loadStatus';
import NavigationService from 'hkufui/src/NavigationService';
import { deepLink } from 'hkufui/config';
import { encrypt } from 'hkufui/src/safe';

import { fetchPosts } from '../Preview/PostPreviewLoader/loadPosts';
import { onLoad } from '../Post/viewActions';

const locale = localize({ language: 'en', country: 'hk' });

function requestCompose({ payload, alert, link, credential }) {
  return async dispatch => {
    dispatch({ type: REQUEST_POST_CREATE });
    alert(locale['new.creatingPost'], 1000);
    try {
      const response = await fetch(link, {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          ...credential,
          ...payload
        })
      });
      const res = await response.json();
      dispatch({ type: END_POST_CREATE });
      switch (res.status) {
        case 200:
          return { ...res.payload };
        case 204:
          return;
        default:
          // failure
          alert(`${locale['new.error']} ${res.status}`);
      }
    } catch (error) {
      // failure
      alert(locale['new.error']);
      dispatch({ type: END_POST_CREATE });
    }
  };
}

export function onReply({ payload, alert, native }) {
  return async (dispatch, getState) => {
    const { location, credential } = getState();
    const { recentTopic: topicId, courseId } = location;

    await dispatch(requestCompose({
      payload: {
        ...payload,
        ...(!native && { code: courseId })
      },
      alert,
      link: native ? link(reply.native({ topicId })) : link(reply.moodle({ topicId })),
      credential: {
        username: credential.userId,
        token: credential.token,
        ...(!native && { moodleKey: credential.moodleKey })
      }
    }));

    alert(locale['reply.success'], 2000, true);
    // go back twice as <Compose /> + <ComposePreview />
    NavigationService.goBackNTimes(2);
    // refresh replies
    dispatch(onLoad({ id: topicId }));
  };
}

export function onCompose({ payload, alert, native }) {
  return async (dispatch, getState) => {
    const { location, credential } = getState();
    const { courseId } = location;

    const { topicId } = await dispatch(requestCompose({
      payload,
      alert,
      link: native ? link(compose.native({ courseId })) : link(compose.moodle({ courseId })),
      credential: {
        username: credential.userId,
        token: credential.token,
        ...(!native && {moodleKey: credential.moodleKey})
      }
    }));

    alert(locale['new.createdPost'], 2000, true);
    // go back twice as <Compose /> + <ComposePreview />
    NavigationService.goBackNTimes(2);
    if (topicId) {
      // deep link to newly created post
      Linking.openURL(`${deepLink.prefix}${deepLink.post(encrypt(topicId))}`);
    }
    // refresh post list
    dispatch(fetchPosts());
  };
}

export default (state = {}, action = {}) => {
  switch (action.type) {
    case REQUEST_POST_CREATE:
      return {
        ...state,
        status: LOADING
      };
    case END_POST_CREATE:
      return {
        ...state,
        status: STILL
      };
    default:
      return state;
  }
};
