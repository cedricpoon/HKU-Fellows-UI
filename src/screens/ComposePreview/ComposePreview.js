import React, { Component } from 'react';
import { Container } from 'native-base';
import { withNavigation } from 'react-navigation';
import { connect } from 'react-redux';
import { format } from 'timeago.js';
import PropTypes from 'prop-types';

import { Header, PostSwipable } from 'hkufui/components';
import { localize } from 'hkufui/locale';
import { STILL, LOADING } from 'hkufui/src/constants/loadStatus';
import { show } from 'hkufui/src/toastHelper';

import postStyles from '../Post/Styles';
import styles from './Styles';
import { onCompose, onReply } from './createActions';

const locale = localize({ country: 'hk', language: 'en' });
const alert = (message, duration, success = false) => {
  show({ message, duration, hideButton: duration <= 1000, type: success ? 'success' : null });
}

export class ComposePreview extends Component {
  constructor(props) {
    super(props);
    this.state = { title: '', subtitle: null, hashtags: null, content: '', anonymity: false, native: true, reply: false };
    this._composePost = this._composePost.bind(this);
  }

  componentDidMount() {
    const { navigation } = this.props;
    // passed props from <ViewMenu />
    if (navigation) {
      const { params: payload } = navigation.state;
      this.setState({ ...payload });
    }
  }

  _composePost() {
    const { onComposeNative, onComposeMoodle, onReplyNative, onReplyMoodle } = this.props;
    const { title, subtitle, hashtags, content, anonymity, native, reply } = this.state;
    if (reply) {
      if (native)
        onReplyNative({ content, anonymity });
      else
        onReplyMoodle({ content });
    } else {
      if (native)
        onComposeNative({ title, subtitle, hashtags, content, anonymity });
      else
        onComposeMoodle({ title, content });
    }
  }

  render() {
    const { title, subtitle, content, anonymity, native, reply } = this.state;
    const { username, status } = this.props;

    return (
      <Container>
        <Header
          title={{
            context: title,
            numberOfLines: 4,
            color: native ? postStyles.nativeTitle.color : postStyles.moodleTitle.color
          }}
          subtitle={subtitle && {
            context: subtitle,
            numberOfLines: 3
          }}
          animated={false}
          backable={status === STILL}
          rightIcon='send'
          rightStyle={styles.send}
          onRightPress={status === STILL ? this._composePost : null}
        />
        <PostSwipable
          comments={status === STILL ? [{
            id: '0'.repeat(64), /* mock of SHA-256 hash */
            author: anonymity ? null : username,
            timestamp: format(Date.now()),
            content: `*{ ${!reply ? locale['new.previewWordings'] : locale['new.replyWordings']} }*\n\n${content}`,
            temperature: 0 /* no temperature at first */
          }] : null}
          native={native}
        />
      </Container>
    );
  }
}

ComposePreview.propTypes = {
  username: PropTypes.string,
  onComposeNative: PropTypes.func.isRequired,
  onComposeMoodle: PropTypes.func.isRequired,
  onReplyNative: PropTypes.func.isRequired,
  onReplyMoodle: PropTypes.func.isRequired,
  status: PropTypes.oneOf([ STILL, LOADING ])
}

const mapStateToProps = state => ({
  username: state.credential.userId,
  status: state.compose.status
});

const mapDispatchToProps = dispatch => ({
  onComposeNative: ({ title, subtitle, hashtags, content, anonymity }) => {
    dispatch(onCompose({
      payload: {
        title, subtitle, content, anonymous: anonymity ? '1' : '0',
        hashtag: hashtags && encodeURI(JSON.stringify(hashtags))
      },
      alert, native: true
    }));
  },
  onComposeMoodle: ({ title, content }) => {
    dispatch(onCompose({
      payload: { title, content },
      alert, native: false
    }));
  },
  onReplyNative: ({ content, anonymity }) => {
    dispatch(onReply({
      payload: { content, anonymous: anonymity ? '1' : '0' },
      alert, native: true
    }));
  },
  onReplyMoodle: ({ content }) => {
    dispatch(onReply({
      payload: { content },
      alert, native: false
    }));
  }
})

export default withNavigation(connect(
  mapStateToProps,
  mapDispatchToProps
)(ComposePreview));
