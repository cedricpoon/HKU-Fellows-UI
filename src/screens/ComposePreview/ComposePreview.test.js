import React from 'react';
import { shallow } from 'enzyme';
import ComposePreview from './ComposePreview';

describe('Testing ComposePreview Screen', () => {

  it('renders as expected', () => {
    const wrapper = shallow(
      <ComposePreview username='alanchan' />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
