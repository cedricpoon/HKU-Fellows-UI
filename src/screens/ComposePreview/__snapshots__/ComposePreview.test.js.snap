// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ComposePreview Screen is creating post 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withNavigation(Connect(ComposePreview))
    onComposeMoodle={[Function]}
    onComposeNative={[Function]}
    onReplyMoodle={[Function]}
    onReplyNative={[Function]}
    status="LOADING"
    username="alanchan"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": undefined,
        "_currentValue2": undefined,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": undefined,
          "_currentValue2": undefined,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Testing ComposePreview Screen renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withNavigation(Connect(ComposePreview))
    onComposeMoodle={[Function]}
    onComposeNative={[Function]}
    onReplyMoodle={[Function]}
    onReplyNative={[Function]}
    status="STILL"
    username="alanchan"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": undefined,
        "_currentValue2": undefined,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": undefined,
          "_currentValue2": undefined,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
